### ISSUE TITLE
DC-LLP-027: Auto re-training pipeline for Picker models

### ISSUE THREAD

**Issue by radoxtech** [radoxtech/diricode#422]
## DC-LLP-027 — Auto Re-training Pipeline for Picker Models

**Epic**: #3 (Provider Router — Transport Layer)
**Package**: `@diricode/dirirouter` (`packages/dirirouter/`)
**Iteration**: v2.0+
**ADR**: [ADR-049](docs/adr/adr-049-llm-picker.md), [ADR-055](docs/adr/adr-055-diri-router-unified-package.md)
**Previously**: LLM Picker epic #389 (closed) → Sub-Epic #395 (closed)

---

### Goal

Implement an automated pipeline that periodically re-trains the Tier 2 (BERT/MiniLM) and Tier 3 (Tiny LLM) models using collected decision feedback data, improving routing accuracy over time without manual intervention.

### Why This Matters

The Picker cascade learns from every routing decision: which model was selected, whether the user was satisfied, what the quality/speed/cost scores were. Over time, this feedback data can significantly improve classification accuracy and reasoning quality. Without automated re-training, the models become stale as new LLMs are added and usage patterns evolve.

#### Data Collection
- Source: `score_history` table (#408) + `decision_log` table (from proxy #462)
- Training pair format for Tier 2: `{ prompt, task_category, reward }` 
- Training pair format for Tier 3: `{ prompt, constraints, selected_model, reward, reasoning }`
- Minimum samples threshold: 500 decisions before first re-training

#### Pipeline Stages
1. **Extract**: Query SQLite for decisions since last training run
2. **Transform**: Clean data, compute reward signals, split train/validation (80/20)
3. **Fine-tune Tier 2**: Update category centroids (no full re-training needed for MiniLM)
   - Recompute centroid vectors from successful classifications
   - Adjust decision boundaries based on misclassification feedback
4. **Fine-tune Tier 3**: LoRA fine-tuning of quantized model on reasoning examples
   - Only if sufficient high-quality reasoning examples (>100 with explicit feedback)
   - Otherwise: update prompt template with few-shot examples from best decisions
5. **Validate**: Run validation set, compare accuracy against current production model
6. **Promote**: If validation accuracy > current - 2% (regression guard), swap model files
7. **Archive**: Store old model version with training metadata

#### Scheduling
- Trigger: Manual CLI command (`diricode picker retrain`) or scheduled (weekly if >500 new decisions)
- Lock file prevents concurrent training runs
- Progress reported to telemetry sink

### Acceptance Criteria

- [ ] Training pair extraction from SQLite with configurable date range
- [ ] Tier 2 centroid recomputation produces updated `task-centroids.json`
- [ ] Tier 3 prompt template updated with best few-shot examples
- [ ] Validation accuracy comparison: new model vs current production
- [ ] Regression guard: new model must not regress >2% on validation set
- [ ] ONNX model files versioned with training metadata (date, sample count, accuracy)
- [ ] CLI command: `diricode picker retrain [--tier 2|3|all] [--since DATE] [--dry-run]`
- [ ] Dry-run mode shows what would change without modifying model files
- [ ] Unit tests cover: extraction, centroid recomputation, validation comparison, regression guard

### Architecture Notes

- Lives in `packages/dirirouter/src/picker/training/`
- Pipeline runner: `packages/dirirouter/src/picker/training/pipeline.ts`
- Data extractor: `packages/dirirouter/src/picker/training/extractor.ts`
- Model versioning: `packages/dirirouter/models/versions/` directory
- Does NOT require Python — all training in TypeScript/ONNX (centroid recomputation is vector math, LoRA is deferred to v3)

### References

- [ADR-049](docs/adr/adr-049-llm-picker.md) — LLM Picker architecture
- [ADR-055](docs/adr/adr-055-diri-router-unified-package.md) — Unified diri-router package
- #412 — Tier 2 classifier (model being re-trained)
- #413 — Tier 3 reasoning (model being re-trained)
- #408 — Score history data source
- #462 — Proxy decision log data source

### Dependencies

- Depends on: #412 (Tier 2 model), #413 (Tier 3 model), #408 (score history table), #462 (decision log)
- Required by: None (optimization layer, no downstream blockers)

**Comment by radoxtech** [comment]
## Update — Dependency restructured, scope revised

### What changed:
The routing strategy has been revised based on research. Key impacts on this issue:

1. **#413 (TinyLLM Router) is CLOSED** — Tier 3 re-training is removed from scope
2. **#412 (BERT Classifier) moved to Phase 2** — blocked on training data
3. **New data source**: DC-DR-041 (Routing Telemetry Collector) replaces hypothetical `score_history` + `decision_log`

### Revised scope:
- **Remove**: All Tier 3 (TinyLLM) re-training references
- **Update**: Data source from `score_history` → DC-DR-041 telemetry output
- **Defer**: LoRA fine-tuning (was speculative, no model to fine-tune in Phase 1)
- **Keep**: Tier 2 centroid recomputation, validation pipeline, regression guard, CLI command

### Updated dependency chain:
```
Before: #412 (Tier 2) + #413 (Tier 3) + #408 + #462 → #422 (this)
After:  DC-DR-041 (Telemetry) + #412 (Tier 2, Phase 2) → #422 (this, Phase 3)
```

### Prerequisites:
- [ ] DC-DR-041 collecting telemetry
- [ ] #412 (BERT) implemented with initial model
- [ ] Minimum 500 labeled decisions in telemetry store

**Status**: Remains open, but moved to Phase 3 (after BERT is operational).

### Related closures:
- #413 (TinyLLM) — CLOSED
- #641 — CLOSED, superseded
